### PR TITLE
[compiler] Collect temporaries and optional chains from inner functions

### PR DESCRIPTION
**Pull request by mofeiZ**
Recursively collect identifier / property loads and optional chains from inner functions. This PR is in preparation for #31200

Previously, we only did this in `collectHoistablePropertyLoads` to understand hoistable property loads from inner functions.
1. collectTemporariesSidemap
2. collectOptionalChainSidemap
3. collectHoistablePropertyLoads
    - ^ this recursively calls `collectTemporariesSidemap`, `collectOptionalChainSidemap`, and `collectOptionalChainSidemap` on inner functions
4. collectDependencies

Now, we have
1. collectTemporariesSidemap
    - recursively record identifiers in inner functions. Note that we track all temporaries in the same map as `IdentifierIds` are currently unique across functions
2. collectOptionalChainSidemap
    - recursively records optional chain sidemaps in inner functions
3. collectHoistablePropertyLoads
    - (unchanged, except to remove recursive collection of temporaries)
4. collectDependencies
    - unchanged: to be modified to recursively collect dependencies in next PR

'
---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/facebook/react/pull/31346).
* #31202
* #31203
* #31201
* #31200
* __->__ #31346
* #31199

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts [hunk 1]
```diff
@@ -360,12 +414,23 @@ class Context {
...
+   *   - inner functions cannot be split by scope boundaries and are guaranteed
+   *     to consume their own declarations
+   *   - reassignments within inner functions are tracked as context variables,
+   *     which already have extended mutable ranges to account for reassignments
+   *   - *most importantly* it's currently simply incorrect to compare inner
+   *     function instruction ids (tracked by `decl`) with outer ones (as stored
+   *     by root identifier mutable ranges).
    */
   declare(identifier: Identifier, decl: Decl): void {
+    if (this.inInnerFn) return;
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts:433]: why don't we need to track declarations here?

**Review by josephsavona — APPROVED** [review]
Yesss i wondered about the inner function traversals. I had hacked on explicitly hoisting deps and needed to add the traversal to the temporary sidemap creation too.

File: compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts (hunk 1 above)
Comment (mofeiZ) [reply compiler/packages/babel-plugin-react-compiler/src/HIR/PropagateScopeDependenciesHIR.ts:433]: See the above comment for details, but tldr:
- inner functions cannot produce scope decls by construction
- inner function instruction ids are invalid to compare against outer instruction ids (which we use to determine mutable range eligibility)

**Comment by mofeiZ** [comment]
Minor cleanup: changed `processedInstrsInOptional`'s type from `Map<HIRFunction, Set<InstructionId>>` to `Set<Instruction | Terminal>` (similar to what we do with `InstructionValue` in `InferReferenceEffect`). This is now a lot cleaner

**Pull request merged**